packages/@atjson/renderer-react/src/index.ts

```diff
@@ -73,20 +72,32 @@ export default class ReactRenderer extends Renderer {
   ): Iterator<void, ReactElement | ReactElement[], ReactElement[]> {
     this.renderSubdocuments(annotation);
 
-    let AnnotationComponent =
-      this.componentLookup[annotation.type] ||
-      this.componentLookup[classify(annotation.type)];
+    const annotationChildren = yield;
+    const key = `${annotation.id}-${annotation.start}`;
 
-    if (AnnotationComponent) {
-      return React.createElement(
-        AnnotationComponent,
-        annotation.attributes,
-        ...(yield)
-      );
-    } else {
-      // console.warn(`No component found for "${node.type}"- content will be yielded`);
-      let components = yield;
-      return components;
-    }
+    return React.createElement(ReactRendererConsumer, {
+      key,
+      children: (componentMap: ComponentMap) => {
+        if (Object.keys(componentMap).length === 0) {
```

Comment:
This is an O(n) operation for every annotation render call which is maybe not great! 😬 Might try out a different solution here

Reply:
I think we can possibly assert that there is / isn't a context object (or use a default reference value), possibly something like:

```ts
const EMPTY_COMPONENT_MAP = {};
const ReactRendererContext = React.createContext<ComponentMap>(EMPTY_COMPONENT_MAP);

if (componentMap === EMPTY_COMPONENT_MAP) {
  throw new Error();
}
```

Reply:
Oh nice, I like that!